### PR TITLE
Add Python UDF for Primitive Types

### PR DESCRIPTION
**Pull request by mxwli**
Here's how it's to be used, as well as an explanation of some of its features
```py
import kuzu
db = kuzu.Database("db")
conn = kuzu.Connection(db)

def aplusb(a: int, b: int) -> int:
    return a + b

conn.create_function(name="aplusb", udf=aplusb)
# we can use annotations to help bind the function

print(conn.execute("return aplusb(1, 2)").get_next()[0])
# 0

def atimesb(a, b):
    return a*b

conn.create_function(name="atimesb", udf=atimesb, parameters=[kuzu.Type.INT64, kuzu.Type.INT64], return_type=kuzu.Type.INT64)
# we can explicitly state the parameters and the return type

print(conn.execute("return atimesb(5, 10)").get_next()[0])
# 50

def vector3dDistance(x1: float, y1: float, z1: float, x2: float, y2: float, z2: float) -> float:
    return ( (x1-x2)**2 + (y1-y2)**2 + (z1-z2)**2 ) ** 0.5
    
conn.create_function("v3dDist", vector3dDistance)
# unlike with the C++ UDF, we can specify as many parameters as we want for python UDFs

print(conn.execute("return v3dDist(-100, 1, 2, 200, -5, 10)").get_next()[0])
# just over 300
```
Nested types haven't been implemented yet, but it's been decided to be done.

### CODE REVIEW

File: tools/python_api/src_cpp/py_udf.cpp [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+            throw RuntimeException("Nested type UDFs not implemented yet");
+        }
+    }
+    if (LogicalTypeUtils::isNested(pySignature.retval)) {
+        throw RuntimeException("Nested type UDFs not implemented yet");
+    }
+
+    function_set definitions;
+    definitions.push_back(std::make_unique<ScalarFunction>(name, pySignature.params,
+        pySignature.retval, getUDFExecFunc(udf), getUDFSelectFunc(udf), nullptr /* bindFunc */));
```
Comment (acquamarin) [inline tools/python_api/src_cpp/py_udf.cpp:176]: Some functions don't have select function, (e.g. arithmetic operation +). I think you can skip udfselectfunc() for now.

File: tools/python_api/src_py/connection.py [hunk 2]
```diff
@@ -194,3 +194,24 @@ def set_query_timeout(self, timeout_in_ms: int) -> None:
         """
         self.init_connection()
         self._connection.set_query_timeout(timeout_in_ms)
+
+    def set_udf(self, name: str, udf: Callable[[...], Any], parameters: list[Type] = [], return_type: Type = None) -> None:
```
Comment (acquamarin) [inline tools/python_api/src_py/connection.py:198]: Can you rename to createFunction

**Review by acquamarin — APPROVED** [review]
What is going to happen if a pyhton function has optional arugments or position arguments? How are we going to use them in kuzu?

File: tools/python_api/src_cpp/py_udf.cpp [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+    return result;
+}
+
+static LogicalTypeID pyTypeToLogicalTypeID(const py::handle& ele) {
+    auto datetime_datetime = importCache->datetime.datetime();
+    auto time_delta = importCache->datetime.timedelta();
+    auto datetime_date = importCache->datetime.date();
+    auto uuid = importCache->uuid.UUID();
+    auto inspect_empty = importCache->inspect._empty();
+    if (ele.is(py::type::of(py::none())) || ele.is(inspect_empty)) {
```
[the remark this answers is not in the document]
Comment (manh9203) [reply tools/python_api/src_cpp/py_udf.cpp:37]: I checked and these two functions work in different ways, so I decided to keep it.
- ```pyLogicalType``` takes a Python value and returns its ```LogicalType```, like ```type()``` in Python.
- ```pyTypeToLogicalTypeID``` takes a ```type``` object in Python and returns ```LogicalTypeID``` corresponding to it. It's quite similar to how we use ```is_same()``` to check for a type in C++.

**Comment by manh9203** [comment]
> What is going to happen if a python function has optional arguments or position arguments? How are we going to use them in kuzu?

@acquamarin If we allow Python UDF to have optional or position arguments, we cannot match the right function based on user's parameters with our function framework. So we'll require users to provide all parameters for Python UDF by now.

I haven't read much into how Python handles optional arguments, @mxwli could take a look at this case when he gets back to work.

**Comment by mxwli** [comment]
> > What is going to happen if a python function has optional arguments or position arguments? How are we going to use them in kuzu?
> 
> @acquamarin If we allow Python UDF to have optional or position arguments, we cannot match the right function based on user's parameters with our function framework. So we'll require users to provide all parameters for Python UDF by now.
> 
> I haven't read much into how Python handles optional arguments, @mxwli could take a look at this case when he gets back to work.

Optional arguments and keyword arguments don't have a very good equivalent in Kuzu, so it's not handled at the moment.

**Pull request merged**